Running devfile integration tests on kubernetes cluster (#3233)
openshift/odo: docs/dev/development.adoc

=== Integration and e2e tests

*Prerequisites for Openshift cluster:*

* A `minishift` or OpenShift environment with Service Catalog enabled:
+
----
$ MINISHIFT_ENABLE_EXPERIMENTAL=y minishift start --extra-clusterup-flags "--enable=*,service-catalog,automation-service-broker,template-service-broker"
----

*Prerequisites for Kubernetes cluster:*

* A `kubernetes` environment set up with single node cluster:
+

> For `kubernetes` single node cluster install link:https://kubernetes.io/docs/tasks/tools/install-minikube/[`Minikube`]

Reviewer: `for a single node kubernetes cluster` and only `install` no need for link

Author: Done

$ TEST_EXEC_NODES=1 make test-cmd-cmp
----

> NOTE: To see the number of available integration test file for validation, press `tab` just after writing `make test-cmd-`. However there is a test file `generic_test.go` which handles certain test specs easily and we can run it parallelly by calling `make test-generic`. By calling `make test-integration`, the whole suite will run all the spec in parallel on two ginkgo test node except `service` and `link` irrespective of service catalog status in the cluster. However `make test-integration-service-catalog` runs all specs of service and link tests in parallel on a cluster having service catalog enabled. `make test-odo-login-e2e` doesn't honour environment variable `TEST_EXEC_NODES`. So by default it runs login and logout command integration test suites on a single ginkgo test node sequentially to avoid race conditions during a parallel run.

Reviewer: all the specs

> NOTE: To see the number of available integration test file for validation, press `tab` just after writing `make test-cmd-`. However there is a test file `generic_test.go` which handles certain test spec easily and can run the spec in parallel by calling `make test-generic`. By calling `make test-integration`, the whole suite can run all the spec in parallel on two ginkgo test node except `service` and `link` irrespective of service catalog status in the cluster. However `make test-integration-service-catalog` runs all spec of service and link tests successfully in parallel on cluster having service catalog enabled. `make test-odo-login-e2e` doesn't honour environment variable `TEST_EXEC_NODES`. So by default it runs login and logout command integration test suite on a single ginkgo test node sequentially to avoid race conditions in a parallel run.

Reviewer: `However there is a test file `generic_test.go` which handles certain test specs easily and we can run it parallelly by calling `make test-generic`. By calling `make test-integration`, the whole suite will run all the specs`

`However `make test-integration-service-catalog` runs all the specs of service and link tests in parallel on a cluster having service catalog enabled. `make test-odo-login-e2e` doesn't honor environment variable `TEST_EXEC_NODES`. So by default it runs login and logout command integration test suites on a single ginkgo test node sequentially to avoid race conditions during a parallel run.

Author: Addressed all the review comment.

*Running integration tests on Kubernetes:*

> By default, the link:https://github.com/openshift/odo/tree/master/tests/integration/devfile[`tests`] in `odo` which are under experimental mode run against `kubernetes` cluster. For more information on Experimental mode, please read link:https://github.com/openshift/odo/blob/master/docs/dev/experimental-mode.adoc:[`odo experimental mode`] document.

Author: 👍

> The tests are run against the `odo` binary placed in the PATH which is created by command `make`. Integration tests can be run in two (parallel and sequential) ways. To control the parallel run use environment variable `TEST_EXEC_NODES`. For example devfile tests can be run

Reviewer: `can be run in two ways: parallel and sequential`

> * To run the tests on `kubernetes` cluster:

Reviewer 2: ```suggestion
* To run the tests on Kubernetes cluster:
```
No need for code blocks here.

Author: Done

+

> Enable `KUBERNETES` environment variable

Reviewer 2: ```suggestion
Set the `KUBERNETES` environment variable
```

Author: Done

+
----
$ export KUBERNETES=true
----

+

> Enable `Experimental` mode

Reviewer 2: ```suggestion
Enable the experimental mode
```
Again no need of code block here.

Author: Done

+
----
$ export ODO_EXPERIMENTAL=true
----
+
OR
+
----
$ odo preference set Experimental true -f
----

* To run the test in parallel, on a test cluster (By default the test will run in parallel on two ginkgo test node):

+
Run catalog command integration tests
+
----
$ make test-cmd-devfile-catalog
----
+

* To run the catalog command integration tests sequentially or on single ginkgo test node:
+
Run catalog command integration tests
+
----
$ TEST_EXEC_NODES=1 make test-cmd-devfile-catalog
----

> NOTE: To see the number of available integration test file for validation, press `tab` just after writing `make test-cmd-devfile-`. By calling `make test-integration-devfile`, the whole suite can run all the spec in parallel on two ginkgo test node.

Reviewer 2: ```suggestion
NOTE: To see the number of available integration test files for validation, press `Tab` key just after writing `make test-cmd-devfile-`. By calling `make test-integration-devfile`, the suite will run all test specs in parallel on two ginkgo test nodes.
```

Author: Thanks for suggestion. Done 👍
